lib/ukvmem/vma_file.c

```diff
@@ -0,0 +1,234 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <stddef.h>
+#include <errno.h>
+
+#include "vmem.h"
+
+#include <uk/config.h>
+#include <uk/assert.h>
+#include <uk/falloc.h>
+#include <uk/arch/limits.h>
+#include <uk/arch/paging.h>
+#ifdef CONFIG_HAVE_PAGING
+#include <uk/plat/paging.h>
+#endif /* CONFIG_HAVE_PAGING */
+#include <vfscore/file.h>
+#include <vfscore/vnode.h>
+#include <vfscore/uio.h>
+#include <uk/isr/string.h>
+
+#ifdef CONFIG_LIBUKVMEM_FILE_BASE
+static __vaddr_t vma_op_file_get_base(struct uk_vas *vas __unused,
+				      void *data __unused,
+				      unsigned long flags __unused)
+{
+	return CONFIG_LIBUKVMEM_FILE_BASE;
+}
+#endif /* CONFIG_LIBUKVMEM_FILE_BASE */
+
+int vma_op_file_new(struct uk_vas *vas, __vaddr_t vaddr __unused,
+		    __sz len __unused, void *data, unsigned long *flags,
+		    struct uk_vma **vma)
+{
+	struct uk_vma_file_args *args = (struct uk_vma_file_args *)data;
+	struct uk_vma_file *vma_file;
+
+	UK_ASSERT(data);
+	UK_ASSERT(args->fd >= 0);
+	UK_ASSERT(args->offset >= 0);
+	UK_ASSERT(PAGE_ALIGNED(args->offset));
+
+	/* Shared mappings are not supported */
+	if (*flags & UK_VMA_FILE_SHARED)
+		return -ENOTSUP;
+
+	/* Since we cannot do ISR-safe file accesses in the fault handler,
+	 * we enforce full load at mapping time for now.
+	 *
+	 * TODO: Remove this restriction if possible.
```

Comment:
We may have to implement some kind of uninterruptible sleep for threads and queue a softirq-like task, but that should be possible. (Okay and we have to have some way to prevent going back to the interrupted thread)

Reply:
Yes, in the end we should have a solution for this. Maybe, for this simple file mapping implementation, it is the easiest solution to just save the registers in case we need to call into the file system. Or we keep it as-is and do not support on-demand paging with the simple file mapping. I would favor this one.